src/coreclr/jit/optimizer.cpp

```diff
@@ -5671,20 +5726,20 @@ Compiler::fgWalkResult Compiler::optIsVarAssgCB(GenTree** pTree, fgWalkData* dat
         }
         else if (destOper == GT_LCL_FLD)
         {
-            /* We can't track every field of every var. Moreover, indirections
-               may access different parts of the var as different (but
-               overlapping) fields. So just treat them as indirect accesses */
-
+            // We can't track every field of every var. Moreover, indirections
+            // may access different parts of the var as different (but
+            // overlapping) fields. So just treat them as indirect accesses
+            //
             // unsigned    lclNum = dest->AsLclFld()->GetLclNum();
             // noway_assert(lvaTable[lclNum].lvAddrTaken);
-
+            //
             varRefKinds refs = varTypeIsGC(tree->TypeGet()) ? VR_IND_REF : VR_IND_SCL;
             desc->ivaMaskInd = varRefKinds(desc->ivaMaskInd | refs);
```

Comment:
Would it be out of the scope for this change to also fix cases where "treating `LCL_FLD` conservatively as indirect access" is not enough, and where we can see `OBJ/BLK`-based stores?

<details>
<summary>Test cases that expose the two issues</summary>

```cs
using System;
using System.Runtime.CompilerServices;

unsafe class LoopLocalInvariance
{
    private const int ArrLen = 10;

    public static int Main()
    {
        int[] arr = new int[ArrLen];

        try
        {
            ProblemWithBlkAsg(arr, ArrLen);
            return 101;
        }
        catch (IndexOutOfRangeException) { }

        try
        {
            ProblemWithLclFldAsg(arr, ArrLen);
            return 102;
        }
        catch (IndexOutOfRangeException) { }

        return 100;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void ProblemWithBlkAsg(int[] arr, int idx)
    {
        for (int i = 0; i < ArrLen; i++)
        {
            Unsafe.InitBlock(&i, (byte)idx, 1);
            arr[i] = 0;
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void ProblemWithLclFldAsg(int[] arr, int idx)
    {
        for (int i = 0; i < ArrLen; i++)
        {
            *(byte*)&i = (byte)idx;
            arr[i] = 0;
        }
    }
}
```
</details>

Reply:
I'll take a look, sure.

I have follow on changes where I want to enable more cloning (an in more cases) so would be good to fix all the issues we know of with the analysis.